lib/DotenvPlugin.js

```diff
@@ -0,0 +1,429 @@
+		} else {
+			result = result.replace(template, defaultValue);
+		}
+
+		// if the result equaled what was in process.env and runningParsed then stop expanding
+		if (result === runningParsed[key || ""]) {
+			break;
+		}
+
+		regex.lastIndex = 0; // reset regex search position to re-evaluate after each replacement
+	}
+
+	return result;
+}
+
+/**
+ * Expand environment variables in parsed object
+ * ported from https://github.com/motdotla/dotenv-expand
+ * @param {{ parsed: Record<string, string>, processEnv?: Record<string, string | undefined> }} options expand options
+ * @returns {{ parsed: Record<string, string> }} expanded options
+ */
+function expand(options) {
+	// for use with progressive expansion
+	const runningParsed = /** @type {Record<string, string>} */ ({});
+
+	let processEnv = process.env;
+	if (
+		options &&
+		options.processEnv !== null &&
+		options.processEnv !== undefined
+	) {
+		processEnv = options.processEnv;
+	}
+
+	// dotenv.config() ran before this so the assumption is process.env has already been set
+	for (const key in options.parsed) {
+		let value = options.parsed[key];
+
+		// short-circuit scenario: process.env was already set prior to the file value
+		value =
+			processEnv[key] && processEnv[key] !== value
+				? /** @type {string} */ (processEnv[key])
+				: expandValue(value, processEnv, runningParsed);
+
+		options.parsed[key] = _resolveEscapeSequences(value);
+
+		// for use with progressive expansion
+		runningParsed[key] = _resolveEscapeSequences(value);
+	}
+
+	for (const processKey in options.parsed) {
+		if (processEnv) {
+			processEnv[processKey] = options.parsed[processKey];
+		}
+	}
+
+	return options;
+}
+
+/**
+ * Resolve and validate env prefixes
+ * Similar to Vite's resolveEnvPrefix
+ * @param {string | string[] | undefined} rawPrefix raw prefix option
+ * @returns {string[]} normalized prefixes array
+ */
+const resolveEnvPrefix = (rawPrefix) => {
+	const prefixes = Array.isArray(rawPrefix)
+		? rawPrefix
+		: [rawPrefix || "WEBPACK_"];
+
+	// Check for empty prefix (security issue like Vite does)
+	if (prefixes.includes("")) {
+		throw new Error(
+			"prefix option contains value '', which could lead to unexpected exposure of sensitive information."
+		);
+	}
+
+	return prefixes;
+};
+
+/**
+ * Get list of env files to load based on mode
+ * Similar to Vite's getEnvFilesForMode
+ * @param {InputFileSystem} inputFileSystem the input file system
+ * @param {string} dir the directory containing .env files
+ * @param {string | undefined} mode the mode (e.g., 'production', 'development')
+ * @returns {string[]} array of file paths to load
+ */
+const getEnvFilesForMode = (inputFileSystem, dir, mode) => {
+	if (dir) {
+		return [
+			/** default file */ ".env",
+			/** local file */ ".env.local",
+			/** mode file */ `.env.${mode}`,
+			/** mode local file */ `.env.${mode}.local`
+		].map((file) => join(inputFileSystem, dir, file));
+	}
+
+	return [];
+};
```

Comment:
I think we will need an option for this, some developers can use non standard filenames (I saw it many times), so what about `template` option? It will be array, like `template: [".env", ".env.local", ".env.[mode]", ".env.[mode].local"]`, only `[mode]` placeholder supported